### ISSUE TITLE
Prevent Grunt from stopping the process when a test fails

### ISSUE THREAD

**Issue by teddyhwang** [aeh/grunt-mocha-protractor#4]
Is there a way to have some sort of keepalive option here? Right now, if a test fails, the entire Grunt process stops while the selenium session is still running.

**Comment by aeh** [comment]
The short answer is that there isn't a lot that I can do... but you can.  The tests are most likely failing due to your expect lib (or something else in your test function) throwing an exception.  If you make sure you wrap anything in your tests that can throw an exception in a try/catch and make sure the done function is executed, with or without an error then everything should work.  See http://stackoverflow.com/questions/11235815/is-there-a-way-to-get-chai-working-with-asynchronous-mocha-tests.

A slightly longer answer is that exceptions don't play nicely with asynchronous functions.  You can't easily add a try/catch block around something that will throw and exception after the current code function has finished.  Consider...

``` javascript
function async(callback) {
  process.nextTick(function() {
    throw new Error('Something went wrong');
  }
}

try {
  async(function() {
    console.log('All done!');
  });
} catch (e) {
  console.log('This will never happen');
}
```

The try/catch block has executed and finished before the error in the async function is thrown and so it is never caught.  It will emit an `uncaughtException` event which can be used for a basic console.log etc but wouldn't work so well with closing the session as it ends the process as soon as it is done (ie. no async will work there).

Basically, as far as I am aware anyway, as soon as something in your tests goes async then I can't do much about cleaning up if it throws an exception.

**Comment by aeh** [comment]
Actually that isn't entirely true.  I can add a domain around the run call and catch any error events. However so far that only seems to work for basic async tests like setTimeout.  Anything inside a `ptor.something.then` function doesn't get caught... which kind of defeats the purpose

**Comment by teddyhwang** [comment]
Thanks @aeh - I found that the best way to use Protractor is to use it with [Chai as Promised](https://github.com/domenic/chai-as-promised) and [Mocha as Promised](https://github.com/domenic/mocha-as-promised). Since both Web Driver and Protractor heavily rely on promises, these tools seem to complement each other the best.

**Comment by aeh** [comment]
For what it is worth I have found and added the magic incantation to catch uncaught exceptions in webdriver promises as well now.
